src/lem/constrainer.rs

```diff
@@ -0,0 +1,346 @@
+            let alloc_ptr = Self::allocate_ptr(
+                &mut cs.namespace(|| format!("allocate input {}", &self.input[1])),
+                &store.hydrate_ptr(&witness.input[1])?,
+                &self.input[1],
+            )?;
+            alloc_ptrs.insert(&self.input[1], alloc_ptr.clone());
+            Self::inputize_ptr(cs, &alloc_ptr, &self.input[1])?;
+        }
+
+        // allocate third input
+        {
+            if alloc_ptrs.contains_key(&self.input[2]) {
+                return Err(format!("{} already allocated", &self.input[2]));
+            }
+            let alloc_ptr = Self::allocate_ptr(
+                &mut cs.namespace(|| format!("allocate input {}", &self.input[2])),
+                &store.hydrate_ptr(&witness.input[2])?,
+                &self.input[2],
+            )?;
+            alloc_ptrs.insert(&self.input[2], alloc_ptr.clone());
+            Self::inputize_ptr(cs, &alloc_ptr, &self.input[2])?;
+        }
+
+        let mut num_inputized_outputs = 0;
+
+        let mut alloc_manager = AllocationManager::default();
+
+        let mut stack = vec![(&self.lem_op, None::<Boolean>, String::new())];
+        while let Some((op, branch_path_info, path)) = stack.pop() {
+            match op {
+                LEMOP::MkNull(tgt, tag) => {
+                    if alloc_ptrs.contains_key(tgt.name()) {
+                        return Err(format!("{} already allocated", tgt.name()));
+                    };
+                    let z_ptr = {
+                        if Self::on_concrete_path(&branch_path_info)? {
+                            let Some(ptr) = witness.ptrs.get(tgt.name()) else {
+                                return Err(format!("Couldn't retrieve witness {}", tgt.name()));
+                            };
+                            store.hydrate_ptr(ptr)?
+                        } else {
+                            ZPtr::dummy()
+                        }
+                    };
+                    let alloc_tgt = Self::allocate_ptr(
+                        &mut cs.namespace(|| format!("allocate pointer {}", tgt.name())),
+                        &z_ptr,
+                        tgt.name(),
+                    )?;
+                    alloc_ptrs.insert(tgt.name(), alloc_tgt.clone());
+                    let alloc_tag = alloc_manager.alloc(cs, tag.to_field())?;
+
+                    // If `concrete_path` is Some, then we constrain using "concrete implies ..." logic
+                    if let Some(concrete_path) = branch_path_info {
+                        let Ok(_) = implies_equal(
+                            &mut cs.namespace(|| format!("implies equal for {}'s tag", tgt.name())),
+                            &concrete_path,
+                            alloc_tgt.tag(),
+                            &alloc_tag,
+                        ) else {
+                            return Err(format!("Couldn't enforce implies equal for {}'s tag", tgt.name()))
+                        };
```

Comment:
The form `let value = opt_value.ok_or(Err(format!("Error message")))?;` is more idiomatic, here and elsewhere.